gax/retry_policy.h

```diff
@@ -73,37 +100,55 @@ class LimitedErrorCountRetryPolicy : public RetryPolicy {
     return (!status.IsPermanentFailure() && (failure_count_++) < max_failures_);
   }
 
+  std::chrono::system_clock::time_point OperationDeadline() const override {
+    return c_.now() + rpc_duration_;
+  }
+
  private:
+  Clock c_;
+  std::chrono::milliseconds const rpc_duration_;
   int failure_count_;
   int const max_failures_;
 };
 
 /**
  * Implement a simple "keep trying for this time" retry policy.
  */
-template <typename Clock>
+template <typename Clock = DefaultClock>
 class LimitedDurationRetryPolicy : public RetryPolicy {
  public:
-  template <typename duration_t>
-  LimitedDurationRetryPolicy(duration_t max_duration)
-      : max_duration_(max_duration), deadline_(max_duration_ + Clock::now()) {}
+  template <typename duration1_t, typename duration2_t>
+  LimitedDurationRetryPolicy(duration1_t max_duration, duration2_t rpc_duration,
+                             Clock c = Clock{})
+      : c_(std::move(c)),
+        rpc_duration_(rpc_duration),
+        max_duration_(max_duration),
+        deadline_(max_duration_ + c_.now()) {}
 
   LimitedDurationRetryPolicy(LimitedDurationRetryPolicy const& rhs) noexcept
-      : LimitedDurationRetryPolicy(rhs.max_duration_) {}
+      : LimitedDurationRetryPolicy(rhs.max_duration_, rhs.rpc_duration_,
+                                   rhs.c_) {}
 
   LimitedDurationRetryPolicy(LimitedDurationRetryPolicy&& rhs) noexcept
-      : LimitedDurationRetryPolicy(rhs.max_duration_) {}
+      : LimitedDurationRetryPolicy(rhs.max_duration_, rhs.rpc_duration_,
+                                   std::move(rhs.c_)) {}
 
   std::unique_ptr<RetryPolicy> clone() const override {
     return std::unique_ptr<RetryPolicy>(
         new LimitedDurationRetryPolicy<Clock>(*this));
   }
 
   bool OnFailure(Status const& status) override {
-    return (!status.IsPermanentFailure() && Clock::now() < deadline_);
+    return (!status.IsPermanentFailure() && c_.now() < deadline_);
+  }
+
+  std::chrono::system_clock::time_point OperationDeadline() const override {
+    return c_.now() + rpc_duration_;
```

Comment:
Should this be something like:

```C++
    return std::min(deadline_, c_.now() + rpc_duration_);
```

or rather:

```C++
    // WIN32 defines `min` as a macro, yuck.
    return (std::min)(deadline_, c_.now() + rpc_duration_);
```

Reply:
Talked about this offline with @vam-google a little last week. He can correct me if I'm wrong, but Java clients don't pick the min. Since the 'logical' deadline is usually orders of magnitude larger than the 'rpc' deadline, waiting for the last rpc deadline a little past the logical deadline is fine.

Reply:
As I have mentioned elsewhere, the timeouts (`rpc_duration_` in this code) are typically set to unrealistic values: these RPCs *can* take minutes to execute when the server is under stress or during outages. And in those conditions timing out and retrying is the **opposite** of what you want to do (retrying puts more load and slows down the recovery).

But that is also a discussion for another day, so SGTM.

Reply:
That seems a bit like a Prisoners' Dilemma, where a locally optimal choice (client cancels a slow request and retries it, since it may have gotten lost on the wire or the server barfed on it) can be globally pessimal (server may be overloaded, and cancelled-then-retried requests make the situation worse).

I can't think of a _correct_ thing to do, barring a lower-level out-of-band communication channel with the server.